Support '%i' placeholders for escaping Identifiers

craigfrancis opened the pull request:
Ticket: https://core.trac.wordpress.org/ticket/52506
Change: https://core.trac.wordpress.org/changeset/53575

With WordPress 6.1 ([October 2022](https://make.wordpress.org/core/2022/06/23/wordpress-6-1-planning-roundup/)), the `%i` placeholder will provide a safe way to escape Identifiers (e.g. table/field names).

Technically it's not needed if the variable only contains [permitted characters](https://dev.mysql.com/doc/refman/8.0/en/identifiers.html), is not one of the ~286 [reserved words](https://dev.mysql.com/doc/refman/8.0/en/keywords.html), and is a `literal-string` (a value defined by the developer)... but that's not always possible, or can be hard to prove.

jrfnl commented:
@craigfrancis Please add some tests.

craigfrancis commented:
Yep, sorry Juliette... will need to do later (thought I could make a quick change before going out).

craigfrancis commented:
I've added a few tests, happy to add more, or make other tweaks/changes :-)

jrfnl commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$field = 'post_type';
+$value = 'page';
+
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %s', $wpdb->posts, $field, $value ); // OK.
```
Please also add an "OK" test with `i` placeholders with various different supported modifiers.

I realize that, in practice, it is unlikely that any modifier other than the `[argnum$]` will ever be used, but the sniff should still recognized and handle an `i` placeholder with other modifiers correctly.

craigfrancis replied:
Yep, done :-)

jrfnl replied:
Thanks for that!

jrfnl commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = "%s"', $wpdb->posts, $field, $value ); // QuotedSimplePlaceholder: "Simple placeholders should not be quoted".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE "%i" = %s', $wpdb->posts, $field, $value ); // QuotedSimplePlaceholder: "Simple placeholders should not be quoted".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE `%i` = %s', $wpdb->posts, $field, $value ); // QuotedIdentifierPlaceholder: "Identifier placeholders should not be quoted".
+
```
I imagine a test with `%i`.`%i` (i.e. table name, column name format as used with `join`s) would also be a good addition.

craigfrancis replied:
Yep, done :-)

jrfnl reviewed (changes requested):
Hi @craigfrancis, thank you for this PR and sorry it took me a while before I could review it.

As you can imagine, I wanted to do some research to properly understand the implementation in WP before reviewing this PR as the sniff is a security sniff and we'd want to minimize the risk of false positives/negatives to prevent "teaching" people bad practices.

Thank you for providing the links to help with that research and for all your work on getting the support for placeholders for identifiers into Core.

:point_right: A little more information in the PR description about the implementation details of the feature in WP would have helped to speed up the research, which now took the better part of day. Please keep this in mind for future PRs.

Also note there is a loosely related issue open for the same sniff: #1903

---

Based on my current understanding the `%i` placeholder feature has been implemented in WP as follows:

* The `i` was chosen for the placeholder to minimize the risk of future conflicts with PHP itself as the `i` is basically an alias for the `%d` in C and PHP has chosen to only implement the `d`, so the chances of PHP re-using the `i` for something else are slim.
    Refs:
    - https://core.trac.wordpress.org/ticket/52506#comment:2
    - https://www.php.net/manual/en/function.sprintf.php
    - https://www.tutorialspoint.com/c_standard_library/c_function_sprintf.htm
    - https://3v4l.org/dDYdJ
* The `%i` should not be prequoted (with backticks) within query strings.
* Use of the `%i` with modifiers is supported, just like it is for other placeholders.
* Using `%i` with modifiers, like `%3$i`, will still result in the identifier name being (backtick-)quoted - **_unlike strings and numbers_**, where only "plain" `%s`, `%d` etc will result in a quoted value and use with modifiers will result in a value without wrapping quotes.

Other general notes:
* If quoted, identifiers should always be quoted with backticks (unless the non-default ANSI_QUOTES mode is enabled, then both backticks and double quotes are allowed, but not single quotes). This backtick quoting has been implemented in WP.

My review comments are based on the above outlined understanding.

@craigfrancis Can you confirm that this is a correct summary of the feature implementation in WP ?

---

Now on to the review - I've left most of my notes in-line. Please let me know if you'd like me to clarify something or if you need guidance on how to implement something in a sniff.

Aside from the inline comments:
* The class docblock needs updating to indicate that the `%i` modifier will be supported.
* Probably should also get a @link tag to the ticket you referenced.
* And an extra `@since` tag to indicate that the `%i` placeholder is only supported as of WPCS 3.0.0.

Please also note that the Coding Standards Handbook contains a section about [formatting SQL statements](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#formatting-sql-statements) which mentions placeholders and their correct usage (last paragraph of the section).
I'd recommend for [that section](https://github.com/WordPress/wpcs-docs/blob/dff1d8622d201bf1d9bc3b9896f783a44778b0cc/wordpress-coding-standards/php.md?plain=1#L465) to be updated. An update can be made via a PR to the https://github.com/WordPress/wpcs-docs/ repo.

Aside from the changes which are contained in this PR (and the inline suggestions for expansion on those), I think there is an additional future scope we should possibly consider (doesn't have to be included in this PR - I'd be perfectly happy for us to open an issue about this to be addressed later):

* Add a check to verify that the `%i` placeholders are used only for identifier names.
    - This would need additional research to work out how to reliably distinguish identifier names from values, could possibly, partially be done by looking for an SQL keyword before it, partially by looking for placeholders within backticks.
* Add a check to recommend that the `%i` placeholders are used for identifier names instead of `$wpdb->tablename` or `%1$s`-like placeholders.
    - This would need the same research as mentioned above.
    - This check would need to use the `MinimumWPVersionTrait` to determine whether the recommendation should be shown or not, based on the minimum WP version a plugin/theme supports.

Regarding the version check: keep in mind that WPCS is used by both WP Core as well as plugins and themes and that plugins/themes will often support a range of WP versions. With this in mind, a lot of them will not be able to start using the `%i` modifier in queries until they drop support for WP < 6.1.
_Yes, I realize they can use `has_cap()` and have separate queries for WP < 6.1 and WP 6.1+, but I doubt many plugins will bother with that. Most will just wait with implementing the `%i` modifier until they drop support for WP < 6.1._


I also wonder if we should use the version check for any checks for `%i` in the first place, as `%i` will now no longer be flagged as an `UnsupportedPlaceholder`, while if plugins/themes still support older WP versions, it probably should be. Something like "_The `%i` modifier is only supported in WP 6.1 or higher_". What do you think ?

Either way, hope my remarks help and please let me know if you have any questions.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +409,30 @@ public function process_token( $stackPtr ) {
+			 * Analyse the query for quoted identifier placeholders.
+			 */
+			$regex = '/`%i`/';
+			if ( preg_match_all( $regex, $content, $matches ) > 0 ) {
```
Thanks, yes, I was being lazy and cloned/tweaked the sniff above it.

Because I need to check Simple and Complex Identifier placeholders, I've re-written this.

Just to note, I am only looking for a quote character before - it keeps the RegEx simpler, and I can't think of a quote only appearing immediately after a backtick quoted identifier doing anything other than causing an SQL parse error.

jrfnl replied:
> Because I need to check Simple and Complex Identifier placeholders, I've re-written this.

👍🏻 

> Just to note, I am only looking for a quote character before - it keeps the RegEx simpler, and I can't think of a quote only appearing immediately after a backtick quoted identifier doing anything other than causing an SQL parse error.

That actually has the side-effect of making the error message look unbalanced (not the missing backtick after the placeholder):
```
 243 | ERROR   | Complex placeholders used for Identifiers (%i) in the query string in $wpdb->prepare() are always quoted automagically. Found:
     |         | `%1$'x10.3i. (WordPress.DB.PreparedSQLPlaceholders.QuotedIdentifierPlaceholder)
```

I have left a suggestion inline on how to improve the regex/resulting match use.

craigfrancis replied:
Thanks, I've used your suggestion (more details below).

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -645,7 +661,7 @@ protected function analyse_implode( $implode_token ) {
 		}
 
-		return (bool) preg_match( '`^(["\'])%[dfFs]\1$`', $array_fill_params[3]['raw'] );
+		return (bool) preg_match( '`^(["\'])%[dfFs]\1$`', $array_fill_params[3]['raw'] ); // Values, not Identifiers `%i`.
```
I've added to the function docblock, I'm just wondering though, when I'm being lazy I can skip reading the docblock, but I would notice an inline comment that explicitly tells me not to do a thing on the line that I'm editing... happy to remove, but thought I'd mention before doing so.

craigfrancis replied:
Due to the new `IdentifierWithinIN` error, the inline comment isn't needed any more (someone would need to remove that error if they did try to add `i`).

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +409,30 @@ public function process_token( $stackPtr ) {
 				if ( ! empty( $matches[0] ) ) {
 					foreach ( $matches[0] as $match ) {
-						if ( preg_match( '`%[dfFs]`', $match ) !== 1 ) {
+						if ( preg_match( '`%[dfFsi]`', $match ) !== 1 ) {
```
Yes, you're right, I needed to support complex identifier placeholders as well, thanks for picking that up. Hopefully the updated test covers this.

I've also added your test.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +409,30 @@ public function process_token( $stackPtr ) {
+					$this->phpcsFile->addError(
+						'Identifier placeholders should not be quoted in the query string in $wpdb->prepare(). Found: %s.',
+						$i,
+						'QuotedIdentifierPlaceholder',
```
As I've re-written this test, I've kept `QuotedIdentifierPlaceholder` for now, because it does Simple and Complex placeholder checks (happy to change).

jrfnl replied:
Makes sense to me. 👍🏻

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -392,9 +392,9 @@ public function process_token( $stackPtr ) {
 
 			/*
-			 * Analyse the query for quoted placeholders.
+			 * Analyse the query for quoted simple placeholders.
```
Thanks, I've gone with your suggestion, and added "value" (to cover string/int/float), with a separate line to note that Identifiers are checked separately.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +409,30 @@ public function process_token( $stackPtr ) {
 			}
 
+			/*
+			 * Analyse the query for quoted identifier placeholders.
```
Thanks, updated (but dropped "simple" because it now does both).

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
 $wpdb->query($wpdb->prepare("delete from wp_postmeta where post_id = $target_postId AND meta_key like 'google_snippets'")); // Bad, the PreparedSQL sniff will also kick in and throw an error about `$target_postId`.
+
+$field = 'post_type';
+$value = 'page';
```
Good point, removed (I was checking my tests against WordPress).

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %2$i = %3$s', $wpdb->posts, $field, $value ); // UnquotedComplexPlaceholder: "Complex placeholders used" (not supported with %i, will eventually be disabled with $wpdb->allow_unsafe_unquoted_parameters).
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = "%s"', $wpdb->posts, $field, $value ); // QuotedSimplePlaceholder: "Simple placeholders should not be quoted".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE "%i" = %s', $wpdb->posts, $field, $value ); // QuotedSimplePlaceholder: "Simple placeholders should not be quoted".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE `%i` = %s', $wpdb->posts, $field, $value ); // QuotedIdentifierPlaceholder: "Identifier placeholders should not be quoted".
```
Good point, removed.

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %s', $wpdb->posts, $field, $value ); // OK.
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %s', $wpdb->posts, $value ); // ReplacementsWrongNumber: "Incorrect number of replacements".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %x', $wpdb->posts, $field, $value ); // UnsupportedPlaceholder: "Unsupported placeholder used".
```
Thanks, I've updated the comment... as noted elsewhere, I can remove this test, as it is covered by the test on line 28, but I quite like seeing comparable tests together.

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %s', $wpdb->posts, $field, $value ); // OK.
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %s', $wpdb->posts, $value ); // ReplacementsWrongNumber: "Incorrect number of replacements".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %x', $wpdb->posts, $field, $value ); // UnsupportedPlaceholder: "Unsupported placeholder used".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %2$i = %3$s', $wpdb->posts, $field, $value ); // UnquotedComplexPlaceholder: "Complex placeholders used" (not supported with %i, will eventually be disabled with $wpdb->allow_unsafe_unquoted_parameters).
```
Thanks, I've updated the tests so the majority of them now only use one placeholder.

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %s', $wpdb->posts, $value ); // ReplacementsWrongNumber: "Incorrect number of replacements".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = %x', $wpdb->posts, $field, $value ); // UnsupportedPlaceholder: "Unsupported placeholder used".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %2$i = %3$s', $wpdb->posts, $field, $value ); // UnquotedComplexPlaceholder: "Complex placeholders used" (not supported with %i, will eventually be disabled with $wpdb->allow_unsafe_unquoted_parameters).
+$wpdb->prepare( 'SELECT ID FROM %i WHERE %i = "%s"', $wpdb->posts, $field, $value ); // QuotedSimplePlaceholder: "Simple placeholders should not be quoted".
```
I've replaced these tests, but I do still have a couple of `%s` based tests, like `'WHERE %i = "%s"'` to show them in context (as in, yes, it works here, but not here)... happy to remove, I just find it easier to see the difference in tests when they are next to each other.

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'SELECT ID FROM %i WHERE "%i" = %s', $wpdb->posts, $field, $value ); // QuotedSimplePlaceholder: "Simple placeholders should not be quoted".
+$wpdb->prepare( 'SELECT ID FROM %i WHERE `%i` = %s', $wpdb->posts, $field, $value ); // QuotedIdentifierPlaceholder: "Identifier placeholders should not be quoted".
+
+$fields = ['name_first', 'name_last'];
```
Removed

craigfrancis commented:
Not complete yet... I've got to look at the `array_fill()` test.

---

Thanks @jrfnl, your summary is correct.

To confirm, you're right, Identifiers are always backtick-quoted (`%i`, `%1$i`, `%10i`, etc); and that's on purpose... my long term intention (~10 years) is to deprecate and remove the risky feature where values aren't quoted when modifiers are used, because it relies on every developer correctly adding the quote characters themselves (mistakes are made, WPCS is good at trying to find these, but not everyone uses WPCS).

I want `wpdb::prepare()` to do all the quoting/escaping work for the developer (correctly and consistently), and for the SQL to be provided as a `literal-string` (a developer defined string, which is something that's easier to test)... the next step to support this will be via [#54042](https://core.trac.wordpress.org/ticket/54042) so we can use `$wpdb->prepare('WHERE id IN (%...d)', $ids)`.

---

As to the changes, I've go though the inline ones separately (feel free to resolve/comment as needed), and the rest...

I've updated the class docblock, added the `@link`, and `@since`.

The suggested additional checks, could they be in a separate Issue/PR, as they could be fairly complicated, and might need further discussion (I'm happy to start those as new Issues).

You're right about versioning, I doubt many developers will write 2 versions of the same query based on `has_cap()`... I'm not too familiar with WPCS, does it have a way to change its recommendations based on a version number? maybe `get_wp_version_from_cli()`?

And thank you for your feedback/suggestions, they were really helpful.

craigfrancis commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,24 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+		implode( ',', array_fill( 0, count($fields), '%i' ) )
+	),
+	$fields
+); // ReplacementsWrongNumber: "Incorrect number of replacements" (`analyse_implode()` shouldn't work with %i, it deals with *values* with `IN()`; a variable number of *identifiers* should come later, e.g. `CONCAT(%...i)`, https://core.trac.wordpress.org/ticket/54042 ).
```
Good point, I've added the error `IdentifierWithinIN`, not sure it's the best place to put it, but at the same time I'm not sure I should really be the one to make larger changes elsewhere.

jrfnl replied:
> I've added the error IdentifierWithinIN, not sure it's the best place to put it

I agree, it is a bit of an odd one out in that function, but to be honest, it does seem like the most straight-forward way to add it, so I'm fine with having it there for now.

> I'm not sure I should really be the one to make larger changes elsewhere.

I understand, but you're doing so well ;-)

craigfrancis commented:
Finishing off from yesterday, I've:

- Added a comment on [#1903](https://github.com/WordPress/WordPress-Coding-Standards/issues/1903) about how the UnquotedComplexPlaceholder warning might not need to be ignored.
- Started [Issue #2078](https://github.com/WordPress/WordPress-Coding-Standards/issues/2078) to verify that the `%i` placeholders are only used for identifiers.
- Started [Issue #2079](https://github.com/WordPress/WordPress-Coding-Standards/issues/2079) to recommend the use of `%i`.
- Started [Issue #103](https://github.com/WordPress/wpcs-docs/issues/103) on the Coding Standards Handbook.

I might need to make some tweaks regarding the recommendations based on the `MinimumWPVersionTrait`... and/or adding a new check that will report "The %i modifier is only supported in WP 6.1 or higher".

jrfnl commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,56 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'WHERE %i = %s', $field, $value ); // OK.
+$wpdb->prepare( 'WHERE %i = %s', $value ); // ReplacementsWrongNumber.
+$wpdb->prepare( 'WHERE %i = %x', $field, $value ); // UnsupportedPlaceholder & ReplacementsWrongNumber.
+$wpdb->prepare( 'WHERE %i = %2$s', $field, $value ); // UnquotedComplexPlaceholder (to be disabled in the future with $wpdb->allow_unsafe_unquoted_parameters).
```
```suggestion
$wpdb->prepare( 'WHERE %i = %2$s', $field, $value ); // UnquotedComplexPlaceholder.
```

(not relevant yet and won't be for quite a while)

craigfrancis replied:
Thanks, updated (and yes, it will be a long time).

jrfnl commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,56 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+$wpdb->prepare( 'WHERE `%1$010.3i` IS NULL', $field ); // QuotedIdentifierPlaceholder.
+$wpdb->prepare( "WHERE `%1$'x10.3i` IS NULL", $field ); // QuotedIdentifierPlaceholder.
+
+// Should throw error for backticks around identifier placeholders.
+// Should NOT throw warning encouraging adding quotes.
```
These comments are not really needed, they were part of the previous review and you fixed what they pointed out (and added the test).

craigfrancis replied:
Thanks, removed.

jrfnl commented on WordPress/Tests/DB/PreparedSQLPlaceholdersUnitTest.inc:
```diff
@@ -207,3 +207,56 @@ $wpdb->query( $wpdb->prepare( "DELETE FROM $wpdb->wp_options WHERE option_name L
+		implode( ',', array_fill( 0, count($fields), '%i' ) )
+	),
+	$fields
+); // IdentifierWithinIN.
```
```suggestion
); // ReplacementsWrongNumber + IdentifierWithinIN.
```

craigfrancis replied:
Thanks, updated.

jrfnl reviewed (changes requested):
Hi @craigfrancis, thank you so much for making all those changes and opening the issues here and in the other repo!

I've gone through the whole PR again and there are only some (mostly small) "dotting the i's, crossing the t's"-kind of remarks I have left. Other than that, I'm happy with this PR to be merged.

Feel free to squash all the commits together when you make those final updates.

I noticed a couple of remarks in the tests about spaces in placeholders not being supported by the regex.
If I look at the regex, it is supported, but only in combination with a width specifier.
IIRC this was based on the PHP spec for it, though it's been a while since I created that regex. If it's not from the PHP spec, then there may be something said about this in the original PR (about why that choice was made).


> The suggested additional checks, could they be in a separate Issue/PR, as they could be fairly complicated, and might need further discussion

Thanks for opening those issues. I'm fine with just having them as issue reminders for now.

> You're right about versioning, I doubt many developers will write 2 versions of the same query based on `has_cap()`... I'm not too familiar with WPCS, does it have a way to change its recommendations based on a version number? maybe `get_wp_version_from_cli()`?

Yes, sort of.
The `MinimumWPVersionTrait` is designed for that, but relies on people either setting a `minimum_supported_version` property in their custom PHPCS ruleset or passing it on the command-line like so: `--runtime-set minimum_supported_wp_version x.x`.

If no version is passed by the WPCS "customer", the trait presumes a version which is three versions behind the last WP major (and yes, I'm aware that the `$minimum_supported_version` in the trait needs updating for that, this will be done just before the release of the next WPCS version to make sure its in line with the reality at that time).

> I might need to make some tweaks regarding the recommendations based on the MinimumWPVersionTrait... and/or adding a new check that will report "The %i modifier is only supported in WP 6.1 or higher".

That would be a great addition! Having said that, I'm happy for that to be handled in a separate PR. No need to let it be a blocker for this one. Looking forward to your next PR :grin:!


And just for the record:
* The tests on lines 214-216, 218 are testing pre-existing functionality and are not directly related to changes made in this PR. I'm fine with having those extra tests though, so no need to change/remove them.
* There also seems to be some redundancy in the tests. Again, I'm fine with that. Rather too many tests than too little. :+1:

craigfrancis commented:
Not complete yet, just updating the comments (thanks for the suggestions).

I need to re-look at the RegEx one, as my original one checked for a quote prefix, but allowed the developer to accidentally forget the end quote... off the top of my head `"%i something"` (admittedly that would probably result in a parse error, but I'll look at it properly later).

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -37,6 +37,7 @@
+ * @link https://core.trac.wordpress.org/changeset/53575
  *
  * @package WPCS\WordPressCodingStandards
  *
```
Thanks, added.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +411,36 @@ public function process_token( $stackPtr ) {
+			/*
+			 * Analyse the query for backtick-quoted identifier placeholders.
+			 *
+			 * @since 3.0.0
```
Thanks, removed.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +411,36 @@ public function process_token( $stackPtr ) {
+						}
+					}
+				}
+				unset( $match, $matches );
```
thanks, I've used your suggestion... I think your RegEx works well, and I've been trying to solve a problem that doesn't exist... if I can come up with an example I'll make a new PR.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -605,6 +629,11 @@ protected function analyse_sprintf( $sprintf_params ) {
+	 * Identifiers (%i) are not supported, as this function is designed to work
+	 * with `IN()`, which contains a list of Values. In the future it should
+	 * be possible to use a variable number of Identifiers, e.g. `CONCAT(%...i)`,
+	 * https://core.trac.wordpress.org/ticket/54042
```
Thanks, updated.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +411,36 @@ public function process_token( $stackPtr ) {
+					foreach ( $matches[0] as $match ) {
+						if ( '%i' !== $match && 'i' === substr( $match, -1 ) ) {
+							$this->phpcsFile->addError(
+								'Complex placeholders used for Identifiers (%%i) in the query string in $wpdb->prepare() are always quoted automagically. Found: %s.',
```
Thanks, updated.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -409,14 +411,36 @@ public function process_token( $stackPtr ) {
 			}
 
+			/*
+			 * Analyse the query for backtick-quoted identifier placeholders.
```
Thanks, updated.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -645,6 +674,15 @@ protected function analyse_implode( $implode_token ) {
 			return false;
 		}
 
+		if ( "'%i'" === $array_fill_params[3]['raw'] ) {
```
Oh, good spot, thanks for that... has been updated.

craigfrancis commented on WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php:
```diff
@@ -645,6 +674,15 @@ protected function analyse_implode( $implode_token ) {
 
+		if ( "'%i'" === $array_fill_params[3]['raw'] ) {
+			$this->phpcsFile->addError(
+				'Cannot use %i for Identifiers within SQL `IN()`.',
```
Thanks, updated.

craigfrancis commented:
I think that's most of the suggestions.

I've not squashed all the commits together yet, in case you want to see what I've changed since.

---

The exception is the use of `MinimumWPVersionTrait`, which probably needs a separate discussion/PR... where I got it partially working with the following diff:

```diff
--- a/WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php
+++ b/WordPress/Sniffs/DB/PreparedSQLPlaceholdersSniff.php
@@ -13,6 +13,7 @@ use PHP_CodeSniffer\Util\Tokens;
 use PHPCSUtils\Utils\PassedParameters;
 use PHPCSUtils\Utils\TextStrings;
 use WordPressCS\WordPress\Helpers\TextStringHelper;
+use WordPressCS\WordPress\Helpers\MinimumWPVersionTrait;
 use WordPressCS\WordPress\Sniff;
 
 /**
@@ -43,9 +44,13 @@ use WordPressCS\WordPress\Sniff;
  *
  * @since 0.14.0
  * @since 3.0.0 Support for the %i placeholder has been added
+ *
+ * @uses  \WordPressCS\WordPress\Helpers\MinimumWPVersionTrait::$minimum_supported_version
  */
 class PreparedSQLPlaceholdersSniff extends Sniff {
 
+	use MinimumWPVersionTrait;
+
 	/**
 	 * These regexes copied from http://php.net/manual/en/function.sprintf.php#93552
 	 * and adjusted for limitations in `$wpdb->prepare()`.
@@ -415,6 +420,7 @@ class PreparedSQLPlaceholdersSniff extends Sniff {
 			/*
 			 * Analyse the query for quoted identifier placeholders.
 			 */
+			if ( version_compare( $this->minimum_supported_version, '6.1', '>=' ) ) {
```

jrfnl commented:
FYI and quite aside from the PR still needing a final review: the patch adding support for `%i` as a placeholder has just been reverted for WP 6.1 as it apparently caused some side-effects for plugins which were clearly not using WPCS and _doing things wrong_ in the first place.

❗ **Until the `%i` patch gets re-committed to WP Core, this PR should be considered "on hold".** ❗ 

For those people now wondering: the "doing it wrong" is having `LIKE '%%%s%%'` in SQL queries provided to `WPDB::prepare()`, which is one of the code patterns this sniff throws warnings/errors for. The correct way would be to have `LIKE %s` in the query with the replacement contains the `%%` delimiters, as is also documented with the function: https://developer.wordpress.org/reference/classes/wpdb/prepare/ (and is exactly what this sniff tells people to do).

craigfrancis commented:
Thanks Juliette, and yep, I'm re-looking at my WP patch for 6.2 (and, on the positive side, I'll be able to include [your suggestions](https://github.com/WordPress/wordpress-develop/pull/3087) this time).

Also, the example can be cut back even further, to "%%%s" - I'd not seen this being used before, I'd have thought the first "%%" would be converted to the literal "%" (as documented), but I didn't think the "%s" (that followed) would be affected by this (it takes the dangerous approach of accepting and escaping the value, but not quoting it, which [the documentation](https://developer.wordpress.org/reference/classes/wpdb/prepare/) says should only happen with "numbered or formatted string placeholders").

craigfrancis commented:
Thanks @jrfnl, I should have some time over the next week.

I've just re-read the comments, and I think my summary on the [Sep 20](https://github.com/WordPress/WordPress-Coding-Standards/pull/2072#issuecomment-1252646455) lists the remaining bits?

- MinimumWPVersionTrait (maybe)
- Squash all the commits together

And a "final review" ([ref Oct 31](https://github.com/WordPress/WordPress-Coding-Standards/pull/2072#issuecomment-1297790999))?

jrfnl commented:
@craigfrancis Just checking - will this PR be ready for a final review over the next few days or not ? I'm working on another PR for this sniff and they are likely to conflict, so I've been holding off on pulling the other PR...

craigfrancis commented:
Sorry @jrfnl, I've not looked at the MinimumWPVersionTrait bit yet (don't know if that's a complicated thing, hopefully not)... but other than that and a squash of the commits, I think it's ready.

craigfrancis commented:
I've added the `MinimumWPVersionTrait`, and squashed commits... I was just checking they were the final things that needed to be done, I can only assume they are correct now, and this patch is ready for the final review.

jrfnl commented:
@craigfrancis I've (finally) done that final review. Turns out I'm really not good at re-reviewing PRs which have been open for a while and have had multiple reviews already.

Either way, to prevent further delays, I've now made some final fixes myself so this PR can get merged. Unfortunately, even though the "Maintainers are allowed to edit this pull request." option is turned on, I get a "permission denied" when I try to push my changes to your branch.

Would you mind giving me push access to your fork of WPCS so I can push that extra commit and merge this PR ?

FYI: This is the commit message for the additional changes:
```
DB/PreparedSQLPlaceholders: final fixes for %i support

* Add sniff to the list of sniffs using the `MinimumWPVersionTrait`
* Fix incorrect implementation of the minimum WP version check.
    This check was never supposed to hide the `QuotedIdentifierPlaceholder` check when the minimum supported WP version is below 6.2.
    Instead, the check is needed to flag the `%i` modifier as an "unsupported placeholder" when used with a minimum supported WP version below 6.2.

Includes tests safeguarding the new `UnsupportedIdentifierPlaceholder` error.
Note: I've left the `phpcs:set WordPress.DB.PreparedSQLPlaceholders minimum_wp_version 6.2` before the original set of test in place to prevent having to up the error/warning nrs for all those tests, while this particular error is safeguarded via separate tests anyway.

Includes a few minor doc tweaks.
Includes a minor efficiency fix (condition order in an `if`).
```

craigfrancis commented:
Hey Juliette, thank you for looking at this, I really appreciate your time looking at this (and everything else you do; my to do list is embarrassing).

I've added you as a collaborator, and happy to make any other changes.

jrfnl commented:
> Hey Juliette, thank you for looking at this, I really appreciate your time looking at this (and everything else you do; my to do list is embarrassing).
> 
> I've added you as a collaborator, and happy to make any other changes.

Thanks for the invite. I've pushed up my changes now. (and I think I've also figured out why that didn't work originally - some branch name mismatch by the looks of it - you may need to delete a stray branch in your fork now)

If you like, have a look at the changes I made - they are in a separate commit so you can easily see what I updated.

Once the CI build has passed, I will merge the PR now.

jrfnl reviewed (approved):
Approving with the inclusion of the final fixes.

craigfrancis commented:
Thanks Juliette :-)

> you may need to delete a stray branch in your fork now

I'll probably delete my branch/fork, considering it's called "patch-1", I suspect I messed up the branching at some point (I daren't look out of fear of embarrassment).

jrfnl commented:
@craigfrancis Thanks for working with me on this and I'm glad we got this in and managed to avoid this PR turning 1 (year old).
